typing/typecore.ml

```diff
@@ -5368,6 +5255,129 @@ and type_let
       | _ -> ()) l;
   (l, new_env, unpacks)
 
+and type_let_def_wrap_warnings
+    ?(check = fun s -> Warnings.Unused_var s)
+    ?(check_strict = fun s -> Warnings.Unused_var_strict s)
+    ~is_recursive ~exp_env ~new_env ~spat_sexp_list ~attrs_list ~pat_list ~pvs
+    type_def =
```

Comment:
I reviewed the second preparatory commit, " separate the detection of unused let-declarations into a wrapper". I believe that the change is correct.

The change moves a part of the logic of `type_let` into an inner higher-order function `type_let_def_wrap_warnings`, with the property that `type_let_def_wrap_warnings` contains a lot of non-trivial code related to warnings but no level management. The level-management code stays in `type_let` and I guess it is easier to refactor this way (with about 120 unrelated lines moved out of the way).

On the other hand, I'm not fond of the refactoring itself: the interface of `type_let_def_wrap_warnings` is fairly complex (10 parameters including a higher-order function), which suggests that it does not cut along a natural logic boundary.

I think that this is okay to have as part of the present level-management PR: `type_let` is a monster function, and a not-so-convincing split does not make the code noticeably worse than before, I just think that there is untapped potential for better refactorings of the code. I am in favor of going as proposed in the context of this PR, and revisiting later.

(My intuition for how to do better would be t tryo, instead of a having a higher-order auxiliary function that takes control of the computation, to define several smaller helper functions that `type_let` would call before and after the tricky type-checking part.)

Reply:
Actually, there is a deeper reason as to why the separation is done this way.
The idea is to separate precisely the unused var warning machinery from the logic of the type inference.
I agree that having many parameters is not nice, but they are almost all required by the warning machinery.
Using a callback in the middle avoids splitting this machinery into incoherent pieces.
This said, there are certainly ways to improve the code. Separation of concerns is the first step.